feat(cadl): support literal type model property

archerzz opened the pull request:
- change json coverter for `InputModelProperty` to set the `DefaultValue` property if the property type is a literal type
  - update `ObjectTypeProperty` so that `DefaultValue` is set if it's from literal type
- change `ModelTypeProviderFields`:
  - exclude literal properties from constructor parameter list
  - literal properties are `internal readonly`
- change `ModelTypeProvider`, update the criteria on properties to be serialized, now literal type property will be serialized regardless it's read-only
- fix a issue in emitter that boolean literal type is not identified
- minor refactoring
- update test projects

resolve #3035

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

ArcturusZhang commented on test/TestProjects/FirstTest-Cadl/Generated/Models/Thing.cs:
```diff
@@ -16,24 +16,27 @@ public partial class Thing
-        /// <summary> Gets or sets the required literal. </summary>
-        public string RequiredLiteral { get; set; }
+        /// <summary> Gets the required literal string. </summary>
+        internal string RequiredLiteralString { get; } = "accept";
```
well in mgmt plane or hlc, this is implemented inside the constructor using "propertyInitializer" like this: https://github.com/Azure/autorest.csharp/blob/80418efe80837f8748227f19ccf32871c12e7391/src/AutoRest.CSharp/Common/Output/Models/Types/SchemaObjectType.cs#L226

but in our implementation, we are not using this way by adding its initialization to the initializer of the corresponding property while building the ctor.

Do we need to align?

archerzz replied:
This is another option. Previously I found initializers which are bound to constructor parameters, which is not feasible.

archerzz replied:
We have the following codes in `ModelWriter`, that's why I choose to use `DefaultValue`:
https://github.com/Azure/autorest.csharp/blob/80418efe80837f8748227f19ccf32871c12e7391/src/AutoRest.CSharp/Common/Generation/Writers/ModelWriter.cs#L211-L219

archerzz replied:
Because for literal types, the values are constants. So maybe that's enough.

ArcturusZhang reviewed (commented):
We have a `InputConstant` in our input types, which is really similar to the literal type.

Do we need to use two things to represent a very similar case?

Currently the `InputConstant` class is only used in `InputParamter`s, but I believe this literal type has very few differences from the LiteralType.

This is what `InputConstant` and `InputLiteralType` looks like:
```csharp
internal record InputConstant(object? Value, InputType Type);
internal record InputLiteralType(string Name, InputType LiteralValueType, object Value, bool IsNullable = false)
```
The `Name` in literal is constant `Literal`, and `IsNullable` makes no sense on literal types (you can only assign one value for it, how could it be null? if it could be null, it only longer only accepts one value).

Therefore I suggest we combine the `InputConstant` and `InputLiteralType` and things would align with how we deal with the same thing about ConstantSchema back using swaggers

archerzz commented:
A couple of changes since last review:
- Modify emitter to add basic support of differentiating integer from float. We'll not try to identify the full list of CADL number types. Instead, we just support `int32` for integer value and `double` for float value.
- Remove literal type properties from generated deserialization codes
- Add E2E test case

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ObjectTypeProperty.cs:
```diff
@@ -167,5 +168,7 @@ private static string CreateExtraPropertyDiscriminatorSummary(CSharpType valueTy
             return updatedDescription;
         }
+
+        public bool ShouldSkipDeserialization => InputModelProperty?.Type is InputLiteralType;
```
I think we can move this into `ModelTypeProvider` and call inside `CreatePropertySerializations`. I wonder maybe we should have two sets of serialization properties - for serialization and for deserialization.

archerzz replied:
Previously I implemented this property in `ObjectTypeProperty` because there are other 2 places (`SerializationBuilder.GetPropertySerializationsFromBag` and `JsonAddtiontalPropertiesSerialization`) which handle `JsonPropertySerliaization`. I rechecked and confirmed that we won't skip deserialization in those 2 places. So I move this check into `ModelTypeProvider`.

I think currently `JsonPropertySerialization` should be sufficient to represent the serialization constraints.

archerzz commented:
@m-nash I've committed change to treat literal type operation parameters as constant parameters (see the last 2nd commit). However, I think it's still better to handle that in the emitter, because:
- By the meaning, it's a constant. It's more like a compiler optimization, v.s. type/name mapping which should be handled in generator.
- Previously it's handled outside generator. When autorest passes in the parameters, their types are already decided.
- We don't have a neat encapsulation on this issue. I have to change a few places in the codes, instead of one line code change in the emitter.
- Another option is to write a Json converter to transform the `Kind` property. But that's almost equal to change in the emitter.
